integration/test/test_progress.py

```diff
@@ -0,0 +1,349 @@
+
+        # Create the test launcher with the above configuration
+        launcher = geopm_test_launcher.TestLauncher(app_conf=app_conf,
+                                                    agent_conf=agent_conf,
+                                                    report_path=cls._report_path,
+                                                    report_signals=report_signals,
+                                                    trace_path=cls._trace_path,
+                                                    trace_signals=trace_signals)
+        launcher.set_num_node(cls._num_node)
+        launcher.set_num_rank(cls._num_rank)
+        # Run the test application
+        launcher.run(cls._test_name)
+
+        # Output to be reused by all tests
+        cls._report = geopmpy.io.RawReport(cls._report_path)
+        cls._trace = geopmpy.io.AppOutput(cls._trace_path + '*')
+
+    def get_hash(self, function_name):
+        """Get hash value from function name.
+
+        Return the hash value associated with the first region name in
+        the report that contains the substring.
+
+        Args:
+            function_name: The substring to match
+
+        Returns:
+            A string of the hex representation of the region hash
+            suitable for matching from the data frame.
+
+        """
+        host = self._report.host_names()[0]
+        region_names = self._report.region_names(host)
+        full_name = [rn for rn in region_names if function_name in rn]
+        if len(full_name) != 1:
+            raise RuntimeError("Single match not found for {}".format(function_name))
+        return hex(self._report.raw_region(host, full_name[0])['hash'])
+
+    def check_progress(self, time, progress, expected_max, msg):
+        """Check a time series of progress values from a region.
+
+        Checks that the progress contains linearly increasing values
+        from zero to expected max.  Assert that a linear fit to the
+        data is a good one.
+
+        Args:
+            time: A sequence of time values
+
+            progress: A sequence of progress values
+
+            expected_max: The expected maximum progress
+
+            msg: Error message to include with any failures
+
+        """
+        epsilon = 0.05
+        abs_epsilon = epsilon * expected_max
+        # Get rid of nan values and values
+        good_idx = numpy.isfinite(progress)
+        max_prog = max(progress[good_idx])
+        util.assertNear(self, expected_max, max_prog, epsilon=epsilon,
+                        msg='{}: maximum value'.format(msg))
+        # Get rid of values at expected_max to avoid bad fit error
+        # when completed threads stop increasing progress
+        good_idx = good_idx & ~numpy.isin(progress, expected_max)
+        self.assertLess(len(progress) * 0.6, sum(good_idx))
+        progress = progress[good_idx]
+        time = time[good_idx]
+        poly_out = numpy.polyfit(time,
```

Comment:
This is fine, but if we have a performance issue in the future there is repeated work in this call that we could eliminate if this is slow.

Reply:
The test runs very quickly, so for now this is not a problem.  If we start using this pattern in tests that use more data, we may want to look into performance improvements.